docs: edit intro testing page (#38006)
angular/angular: aio/content/guide/testing.md

# Testing your application

The Angular CLI downloads and installs everything you need to test an Angular application with the [Jasmine test framework](https://jasmine.github.io/).

> To test an application created using the CLI command [`ng new`](cli/new), run the [`ng test`](cli/test) command:

Reviewer: It feels a little roundabout to say 
> an application created using the CLI command `ng new`

versus 

>an Angular CLI project

## Prerequisites

To get the most out of the Angular testing documentation, you should have a basic understanding of the following concepts:

* HTML
* CSS
* JavaScript
* [Angular fundamentals](guide/architecture)
* [Angular CLI](/cli)

<hr>

> The testing documentation offers tips and techniques for testing Angular applications through a sample application created with the [Angular CLI](cli).

Author: Nope, we decided to get rid of that on Friday and I forgot to push up. And regarding "unit testing" versus "integration testing", etc., it's been really difficult to find external resources for these that are appropriate. If you see anything let me know!

A Chrome browser also opens and displays the test output in the "Jasmine HTML Reporter" as in the following image.

<div class="lightbox">
  <img src='generated/images/guide/testing/initial-jasmine-html-reporter.png' alt="Jasmine HTML Reporter in the browser">
</div>

You can click on a test row to re-run that individual test.
You can also click on a description to re-run the tests in that test group, or "test suite".

The `ng test` command watches for changes.
This means that when you make a change and save, the tests run again, the browser refreshes, and the new test results appear.

## Editing the testing configuration

The CLI sets up a default testing configuration for Jasmine and Karma.
You can edit many testing options by editing the `karma.conf.js` and the `test.ts` files in the `src/` folder.

The `karma.conf.js` file is a partial Karma configuration file.
The CLI constructs the full runtime configuration in memory, based on the application structure as described by the `angular.json` and `karma.conf.js` configuration files.

<div class="callout is-helpful">

<header>Using other test frameworks</header>

> Though the CLI uses the Jasmine test framework by default, you can unit test an Angular application with other testing libraries and test runners.
> This guide specifically covers Jasmine and Karma.
> If you use another option, be sure to research its distinctive installation procedures, configuration, and syntax.

Reviewer: ```suggestion
By default, Angular CLI uses the Jasmine test framework and the Karma test runner. While this
guide only covers this default setup, you can change your project to use other test frameworks
and test runners.
```
I might omit the last sentence here since it feels a bit awkward.

</div>

{@a test-name-location}

## Finding test files

The CLI includes test files when it generates a new Angular application.

> The testing framework identifies test files, or "spec files", by the extension `spec.ts`.

Reviewer: ```suggestion
The testing framework identifies test files, or "spec" files, by the extension `spec.ts`.
```
Would the quotes just be around "spec" here since files parallels the first part of the sentence?

The last line of the log is the most important.
It shows that Karma ran three tests that all passed.

A Chrome browser also opens and displays the test output in the "Jasmine HTML Reporter" as in the following image.

<div class="lightbox">
  <img src='generated/images/guide/testing/initial-jasmine-html-reporter.png' alt="Jasmine HTML Reporter in the browser">
</div>

You can click on a test row to re-run just that test or click on a description to re-run the tests in the selected test group, or "test suite".

Meanwhile, the `ng test` command is watching for changes.

To see this in action, make a small change to `app.component.ts` and save.
The tests run again, the browser refreshes, and the new test results appear.

> ### Editing test configuration

Reviewer 2: Did you mean to adjust the heading level?

```suggestion
## Editing test configuration
```

Are the headings following this one aligned?

Author: Yes, I did mean to change the header level. The assumption in this section is that if you're using the CLI you already have Jasmine Karma configured by default as discussed just above this section. Is it confusing? What do you think? It's still very much a draft mode PR and there's a lot of flexibility. :)

Reviewer 2: You are in a better position to decide this, so let's follow your intuition, @kapunahelewong.

Author: Thanks, @LayZeeDK. :) I thought of you though when I decided to remove some of the intro stuff and bumped this back up to an `h2`! You were right all along! 🌼

The CLI takes care of Jasmine and Karma configuration for you, however you can edit many options by editing the `karma.conf.js` and the `test.ts` files in the `src/` folder.

The `karma.conf.js` file is a partial Karma configuration file.
The CLI constructs the full runtime configuration in memory, based on the application structure according to the `angular.json` file and `karma.conf.js`.

<div class="callout is-helpful">

<header>Using other test frameworks</header>

Though the CLI uses the Jasmine test framework by default, you can unit test an Angular application with other testing libraries and test runners.

> This guide specifically covers Jasmine and Karma, so if you use another option, be sure to research its distinctive installation procedures, configuration, and syntax.

Reviewer 2: More formal suggestion.

```suggestion
This guide specifically covers Jasmine and Karma. If you use another option, be sure to research its distinctive installation procedures, configuration, and syntax.
```

</div>

{@a test-name-location}

## Finding test files

The CLI includes test files when it generates a new Angular application.
You may notice two features of these tests that the CLI generates.

* A test file, or a _spec_ file, extension is `.spec.ts` so that tooling can identify it as a file with tests.
* Spec files for each component are in the same directory as that component.

For example, the CLI gives the `app.component` spec file the extension `.spec.ts` and places this `app.component.spec.ts` in the `src/app` folder with the `AppComponent`.

Notice that the root file name, `app.component`, is the same for both files.
The component is `app.component.ts` and the spec is `app.component.spec.ts`.

Though the file extension `.spec.ts` is a requirement, the location of the tests being in the same directory is the component that they test is a convention.

> This convention makes tests easy to find, helps reveal how a component works in context, and reminds you to include the tests when moving or renaming parts of your application.

Reviewer 2: Is this comma correct and needed?

```suggestion
This convention makes tests easy to find, helps reveal how a component works in context and reminds you to include the tests when moving or renaming parts of your application.
```

Author: Hi @LayZeeDK! You're right! I was thinking of the second part as an independent clause but it's missing the subject. Nice catch. Thanks :)

<hr />

## More info on testing

After you've set up your app for testing, you may find the following testing  guides useful.

> * [Basics of testing components](guide/testing-components-basics)&mdash;discover the basics of testing Angular components.

Reviewer 3: I assume you are planning to arrange the subpages in this order in the TOC, too.

Author: That is correct. We're currently experimenting with the order.

> * [Basics of testing components](guide/testing-components-basics)&mdash;discover the basics of testing Angular components.

Reviewer 3: "discover",  "learn how", "read about", "find out", "uncover" --Should probably standardize descriptions on whatever term you choose.

Perhaps: 
 [Testing components](guide/testing-components-basics): Learn the basics of testing Angular components.
